streams/shared/src/main/scala/zio/stream/ZSink.scala

```diff
@@ -1410,52 +1321,60 @@ object ZSink extends ZSinkPlatformSpecific {
    */
   final def read1[E, A](e: Option[A] => E)(p: A => Boolean): ZSink[Any, E, A, A, A] =
     new SinkPure[E, A, A, A] {
-      type State = Either[E, Option[A]]
+      type State = (Either[E, Option[A]], Chunk[A])
 
-      val initialPure = Step.more(Right(None))
+      val initialPure = (Right(None), Chunk.empty)
 
       def stepPure(state: State, a: A) =
         state match {
-          case Right(Some(_)) => Step.done(state, Chunk(a))
-          case Right(None) =>
-            if (p(a)) Step.done(Right(Some(a)), Chunk.empty)
-            else Step.done(Left(e(Some(a))), Chunk(a))
-          case s => Step.done(s, Chunk(a))
+          case (Right(Some(_)), _) => (state._1, Chunk(a))
+          case (Right(None), _) =>
+            if (p(a)) (Right(Some(a)), Chunk.empty)
+            else (Left(e(Some(a))), Chunk.single(a))
+          case s => (s._1, Chunk.single(a))
         }
 
       def extractPure(state: State) =
         state match {
-          case Right(Some(a)) => Right(a)
-          case Right(None)    => Left(e(None))
-          case Left(e)        => Left(e)
+          case (Right(Some(a)), _) => Right(a)
+          case (Right(None), _)    => Left(e(None))
+          case (Left(e), _)        => Left(e)
+        }
+
+      def cont(state: State) =
+        state match {
+          case (Right(None), _) => true
+          case _                => false
         }
+
+      def leftover(state: State) = state._2
     }
 
   /**
    * Splits strings on newlines. Handles both `\r\n` and `\n`.
    */
   final val splitLines: ZSink[Any, Nothing, String, String, Chunk[String]] =
     new SinkPure[Nothing, String, String, Chunk[String]] {
-      type State = (Chunk[String], Option[String], Boolean)
+      type State = (Chunk[String], Option[String], Boolean, Chunk[String], Boolean)
```

Comment:
How about we replace this with `case class State` instead? Hard to keep track of what is what

Reply:
I did that but I got a weird warning from the Scala compiler, something about not being able to enforce the type at runtime, probably because it is a path dependent type and I don't feel like exporting types that are unnecessary for the user.